File: gempak/source/programs/gd/gdgrib/gdsced.c
#include "gdgrib.h"

void gds_ced ( const float *rnvblk, const int*nnv, int *nbytes,
               unsigned char *cgds, int *iret )
/************************************************************************
 * gds_ced								*
 *									*
 * This subroutine uses the GEMPAK grid navigation for a CED grid to	*
 * generate a GRIB GDS section for this grid.				*
 *									*
 * gds_ced ( rnvblk, nnv, nbytes, cgds, iret )				*
 *									*
 * Input parameters:							*
 *	*rnvblk		const float	GEMPAK grid navigation block	*
 *	*nnv		const int	Size of the navigation block	*
 *									*
 * Input and output parameter:						*
 *	*nbytes		int		Input: # of bytes available in	*
 *					       CGDS			*
 *					Output: # of bytes filled in	*
 *					       CGDS			*
 *									*
 * Output parameters:							*
 *	*cgds		unsigned char	GRIB GDS section		*
 *	*iret		int		Return code			*
 *					  0 = normal return		*
 *					-61 = not enough GDS bytes	*
 *					-62 = # in i is too large	*
 *					-63 = # in j is too large	*
 *					-64 = latitude 1 is bad		*
 *					-65 = longitude 1 is bad	*
 *					-66 = latitude 2 is bad		*
 *					-67 = longitude 2 is bad	*
 *					-68 = rotated CED not supported *
 **									*
 * Log:									*
 * K. Brill/HPC		 8/99						*
 * K. Brill/HPC		 2/00	Document GDS byte numbers		*
 * R. Tian/SAIC		 9/06	Recoded from Fortran			*
 ************************************************************************/
{
    int ibase, kx, ky, ibyts[3], ilat, ilon, ii, idx, nb, ier;
    float rlat1, rlon1, rlat2, rlon2;
/*----------------------------------------------------------------------*/
    *iret = 0;
    ibase = 256;

    if ( *nbytes < 32 ) {
	*iret = -61;
	return;
    }
    if ( ! G_DIFF ( rnvblk[12], 0.0F ) ) {
	*iret = -68;
	return;
    }
    *nbytes = 32;

    /*
     * Get information from the navigation block.
     */
    kx = G_NINT ( rnvblk[4] );
    ky = G_NINT ( rnvblk[5] );
    rlat1 = rnvblk[6];
    rlon1 = rnvblk[7];
    rlat2 = rnvblk[8];
    rlon2 = rnvblk[9];

    /*
     * Start filling up the GDS array.
     */
    idx = 0;
    cgds[idx++] = (unsigned char) (0);
    cgds[idx++] = (unsigned char) (0);
    cgds[idx++] = (unsigned char) (32);
    cgds[idx++] = (unsigned char) (0);
    cgds[idx++] = (unsigned char) (255);
    cgds[idx++] = (unsigned char) (0);

    /*
     * Set bytes 7-8.
     */
    nb = 2;
    gdigit ( &kx, &ibase, &nb, ibyts, &ier );
    if ( ier != 0 ) {
	*iret = -62;
	return;
    }
    for ( ii = 1; ii >= 0; ii-- )  cgds[idx++] = (unsigned char) ( ibyts [ii] );

    /*
     * Set bytes 9-10.
     */
    nb = 2;
    gdigit ( &ky, &ibase, &nb, ibyts, &ier );
    if ( ier != 0 ) {
	*iret = -63;
	return;
    }
    for ( ii = 1; ii >= 0; ii-- )  cgds[idx++] = (unsigned char) ( ibyts [ii] );

    /*
     * Set bytes 11-13.
     */
    ilat = G_ABS ( G_NINT ( rlat1 * 1000. ) );
    nb = 3;
    gdigit ( &ilat, &ibase, &nb, ibyts, &ier );
    if ( ier != 0 ) {
	*iret = -64;
	return;
    }
    if ( rlat1 < 0.0F ) ibyts[2] += 128;
    for ( ii = 2; ii >= 0; ii-- ) cgds[idx++] = (unsigned char) ( ibyts[ii] );

    /*
     * Set bytes 14-16.
     */
    ilon = G_ABS ( G_NINT ( rlon1 * 1000. ) );
    nb = 3;
    gdigit ( &ilon, &ibase, &nb, ibyts, &ier );
    if ( ier != 0 ) {
	*iret = -65;
	return;
    }
    if ( rlon1 < 0.0F ) ibyts[2] += 128;
    for ( ii = 2; ii >= 0; ii-- ) cgds[idx++] = (unsigned char) ( ibyts[ii] );

    /*
     * Set byte 17.
     */
    cgds[idx++] = (unsigned char) (0);

    /*
     * Set bytes 18-20.
     */
    ilat = G_ABS ( G_NINT ( rlat2 * 1000. ) );
    nb = 3;
    gdigit ( &ilat, &ibase, &nb, ibyts, &ier );
    if ( ier != 0 ) {
	*iret = -66;
	return;
    }
    if ( rlat2 < 0 ) ibyts[2] += 128;
    for ( ii = 2; ii >= 0; ii-- ) cgds[idx++] = (unsigned char) ( ibyts[ii] );

    /*
     * Set bytes 21-23.
     */
    ilon = G_ABS ( G_NINT ( rlon2 * 1000. ) );
    nb = 3;
    gdigit ( &ilon, &ibase, &nb, ibyts, &ier );
    if ( ier != 0 ) {
	*iret = -67;
	return;
    }
    if ( rlon2 < 0 ) ibyts[2] += 128;
    for ( ii = 2; ii >= 0; ii-- ) cgds[idx++] = (unsigned char) ( ibyts[ii] );

    /*
     * Set bytes 24-32.
     */
    cgds[idx++] = (unsigned char) (255);
    cgds[idx++] = (unsigned char) (255);
    cgds[idx++] = (unsigned char) (255);
    cgds[idx++] = (unsigned char) (255);
    cgds[idx++] = (unsigned char) (64);
    cgds[idx++] = (unsigned char) (0);
    cgds[idx++] = (unsigned char) (0);
    cgds[idx++] = (unsigned char) (0);
    cgds[idx++] = (unsigned char) (0);

    return;
}

File: gempak/source/programs/na/gd2ndfd/gds_ced.f
	SUBROUTINE GDS_CED  ( rnvblk, nnv, nbytes, cgds, iret )
C************************************************************************
C* GDS_CED								*
C*									*
C* This subroutine uses the GEMPAK grid navigation for a CED grid to	*
C* generate a GRIB GDS section for this grid.				*
C*									*
C* GDS_CED  ( RNVBLK, NNV, NBYTES, CGDS, IRET )				*
C*									*
C* Input parameters:							*
C*	RNVBLK (NNV)	REAL		GEMPAK grid navigation block	*
C*	NNV		INTEGER		Size of the navigation block	*
C*									*
C* Input and output parameter:						*
C*	NBYTES		INTEGER		Input: # of bytes available in	*
C*					       CGDS			*
C*					Output: # of bytes filled in	*
C*					       CGDS			*
C*									*
C* Output parameters:							*
C*	CGDS (NBYTES)	CHAR*1		GRIB GDS section		*
C*	IRET		INTEGER		Return code			*
C*					  0 = normal return		*
C*					-61 = not enough GDS bytes	*
C*					-62 = # in i is too large	*
C*					-63 = # in j is too large	*
C*					-64 = latitude 1 is bad		*
C*					-65 = longitude 1 is bad	*
C*					-66 = latitude 2 is bad		*
C*					-67 = longitude 2 is bad	*
C*					-68 = rotated CED not supported *
C**									*
C* Log:									*
C* K. Brill/HPC		 8/99						*
C* K. Brill/HPC		 2/00	Document GDS byte numbers		*
C************************************************************************
	INCLUDE		'GEMPRM.PRM'
	REAL		rnvblk (nnv)
	CHARACTER*1	cgds (*)
C*
	INTEGER		ibyts (3)
C------------------------------------------------------------------------
	IF ( nbytes .lt. 32 ) THEN
	    iret = -61
	    RETURN
	END IF
	IF ( rnvblk (13) .ne. 0.0 ) THEN
	    iret = -68
	    RETURN
	END IF
	iret = 0
	nbytes = 32
C
C*	Get information from the navigation block.
C
	kx = NINT ( rnvblk (5) )
	ky = NINT ( rnvblk (6) )
	rlat1 = rnvblk (7)
	rlon1 = rnvblk (8)
	rlat2 = rnvblk (9)
	rlon2 = rnvblk (10)
C
C*	Start filling up the GDS array.
C
	cgds (1) = CHAR (0)
	cgds (2) = CHAR (0)
	cgds (3) = CHAR (32)
	cgds (4) = CHAR (0)
	cgds (5) = CHAR (255)
	cgds (6) = CHAR (0)
	ii = 6
C*
	nb = 2
	CALL GDIGIT ( kx, 256, nb, ibyts, ier )
	IF ( ier .ne. 0 ) THEN
	    iret = -62
	    RETURN
	END IF
C*					Set bytes 7-8.
	DO i = 2, 1, -1
	    ii = ii + 1
	    cgds (ii) = CHAR ( ibyts (i) )
	END DO
	nb = 2
	CALL GDIGIT ( ky, 256, nb, ibyts, ier )
	IF ( ier .ne. 0 ) THEN
	    iret = -63
	    RETURN
	END IF
C*					Set bytes 9-10.
	DO i = 2, 1, -1
	    ii = ii + 1
	    cgds (ii) = CHAR ( ibyts (i) )
	END DO
C*
	ilat = IABS ( NINT ( rlat1 * 1000. ) )
	nb = 3
	CALL GDIGIT ( ilat, 256, nb, ibyts, ier )
	IF ( ier .ne. 0 ) THEN
	    iret = -64
	    RETURN
	END IF
	IF ( rlat1 .lt. 0 ) ibyts (3) = ibyts (3) + 128
C
C*					Set bytes 11-13.
	DO i = 3, 1, -1
	    ii = ii + 1
	    cgds (ii) = CHAR ( ibyts (i) )
	END DO
	ilon = IABS ( NINT ( rlon1 * 1000. ) )
	nb = 3
	CALL GDIGIT ( ilon, 256, nb, ibyts, ier )
	IF ( ier .ne. 0 ) THEN
	    iret = -65
	    RETURN
	END IF
	IF ( rlon1 .lt. 0 ) ibyts (3) = ibyts (3) + 128
C
C*					Set bytes 14-16.
	DO i = 3, 1, -1
	    ii = ii + 1
	    cgds (ii) = CHAR ( ibyts (i) )
	END DO
C*					Set byte 17.
	ii = ii + 1
	cgds (ii) = CHAR (0)
C*
	ilat = IABS ( NINT ( rlat2 * 1000. ) )
	nb = 3
	CALL GDIGIT ( ilat, 256, nb, ibyts, ier )
	IF ( ier .ne. 0 ) THEN
	    iret = -66
	    RETURN
	END IF
	IF ( rlat2 .lt. 0 ) ibyts (3) = ibyts (3) + 128
C
C*					Set bytes 18-20.
	DO i = 3, 1, -1
	    ii = ii + 1
	    cgds (ii) = CHAR ( ibyts (i) )
	END DO
	ilon = IABS ( NINT ( rlon2 * 1000. ) )
	nb = 3
	CALL GDIGIT ( ilon, 256, nb, ibyts, ier )
	IF ( ier .ne. 0 ) THEN
	    iret = -67
	    RETURN
	END IF
	IF ( rlon2 .lt. 0 ) ibyts (3) = ibyts (3) + 128
C
C*					Set bytes 21-23.
	DO i = 3, 1, -1
	    ii = ii + 1
	    cgds (ii) = CHAR ( ibyts (i) )
	END DO
	cgds (24) = CHAR (255)
	cgds (25) = CHAR (255)
	cgds (26) = CHAR (255)
	cgds (27) = CHAR (255)
	cgds (28) = CHAR (64)
C*
	cgds (29) = CHAR (0)
	cgds (30) = CHAR (0)
	cgds (31) = CHAR (0)
	cgds (32) = CHAR (0)
C*
	RETURN
	END
